File: algo/61. Rotate List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def rotateRight(self, head, k):
        """
        :type head: ListNode
        :type k: int
        :rtype: ListNode
        """
        if head is None:
            return head
        dummy = ListNode(0)
        dummy.next = head
        cursor = dummy
        size = 0
        while cursor.next != None:
            cursor = cursor.next
            size += 1
        end = cursor
        targetIndex = size - k % size
        cursor = dummy
        while targetIndex > 0:
            cursor = cursor.next
            targetIndex -= 1
        end.next = dummy.next
        newHead = cursor.next
        cursor.next = None
        return newHead
        

File: round3/61. Rotate List.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution {
    public ListNode RotateRight(ListNode head, int k) {
        if (head == null || head.next == null) return head;
        
        ListNode dummy = new ListNode();
        dummy.next = head;
        ListNode cursor = dummy;
        int len = 0;
        while (cursor.next != null) {
            len ++;
            cursor = cursor.next;
        }
        
        cursor.next = head;
        k %= len;
        
        cursor = head;
        ListNode preHead = dummy;
        for (int i = 0; i < len - k; i ++) {
            preHead = preHead.next;
            cursor = cursor.next;
        }
        
        preHead.next = null;
        return cursor;        
    }
}